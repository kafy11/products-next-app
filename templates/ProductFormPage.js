import { useState } from 'react';
import { useRouter } from 'next/router';
import Page from './Page';
import { Button, InputText } from '../components';

const ProductFormPage = ({ title, onSubmit, ...rest }) => {
    const router = useRouter();
    const [name, setName] = useState(rest.name);
    const [price, setPrice] = useState(rest.price);
    const [quantity, setQuantity] = useState(rest.quantity);

    const handleSubmit = (e) => {
        e.preventDefault();
        onSubmit && onSubmit({
            name,
            price,
            quantity
        });
    };

    const handleOnlyNumber = (e) => {
        if([8,9].indexOf(e.keyCode) == -1 && isNaN(e.key)){
            e.preventDefault();
        }
    };

    const handleCancelClick = () => router.push('/');

    return (
        <Page title={title}>
            <form className="mb-3" onSubmit={handleSubmit} >
                <InputText 
                    label="Name" 
                    value={name}
                    onChange={setName}
                    required 
                />

                <InputText 
                    label="Price"
                    value={price}
                    onChange={setPrice}
                    prepend="$"
                    append=".99"
                    onKeyPress={handleOnlyNumber}
                    required
                />

                <InputText
                    label="Quantity" 
                    value={quantity}
                    onChange={setQuantity}
                    onKeyPress={handleOnlyNumber}
                    required
                />

                <Button submit>Submit</Button>
                <Button 
                    color="light"
                    className="ml-2"
                    onClick={handleCancelClick}
                >
                    Cancel
                </Button>
            </form>
        </Page>
    );
}

export default ProductFormPage;