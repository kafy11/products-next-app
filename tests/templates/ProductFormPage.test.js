import { shallow } from 'enzyme';
import ProductFormPage from '../../templates/ProductFormPage';

const mockPush = jest.fn();
jest.mock('next/router', () => ({
    useRouter: () => ({
        push: mockPush
    })
}));

//function to help test preventDefault block on key press
const expectKeyPress = (field, mockEvent, expectedToBlock) => {
    wrapper.find('.product-form-page__'+field).prop('onKeyPress')(mockEvent);
    
    if(expectedToBlock){
        expect(mockEvent.preventDefault).toHaveBeenCalled();
    } else {
        expect(mockEvent.preventDefault).not.toHaveBeenCalled();
    }
};

//function to help test every cases of key press
const expectOnlyNumber = (field) => {
    const testCases = [
        { keyCode: 69, key: 'e', expectedToBlock: true },
        { keyCode: 49, key: '1', expectedToBlock: false },
        { keyCode: 9, key: 'Tab', expectedToBlock: false },
        { keyCode: 8, key: 'Backspace', expectedToBlock: false }
    ];

    testCases.forEach(({ expectedToBlock, ...rest }) => {
        expectKeyPress(field, {
            ...rest,
            preventDefault: jest.fn()
        }, expectedToBlock);
    });
};

//function to help test input value
const expectFieldHasValue = (field, value) => {
    expect(wrapper.find('.product-form-page__'+field).prop('value')).toBe(value);
};

//function to help test input onChange event
const expectChangeValue = (field, value) => {
    wrapper.find('.product-form-page__'+field).prop('onChange')(value);
    expectFieldHasValue(field, value);
};

const props = {
    title: 'Product Form',
    name: 'Test',
    price: '10',
    quantity: '1',
};

let wrapper, handleSubmit;
beforeEach(() => {
    handleSubmit = jest.fn();
    wrapper = shallow(<ProductFormPage {...props} onSubmit={handleSubmit} />);
});

describe('ProductFormPage', () => {
    it('should render correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    describe('name field', () => {
        const field = 'name';
        it('should render the value correctly', () => {
            expectFieldHasValue(field, props.name);
        });

        it('should update value when onChange is called', () => {
            expectChangeValue(field, 'Produto 1');
        });
    });
    
    describe('price field', () => {
        const field = 'price';
        it('should render the value correctly', () => {
            expectFieldHasValue(field, props.price);
        });

        it('should allow only number, tab and backspace', () => {
            expectOnlyNumber(field);
        });

        it('should update value when onChange is called', () => {
            expectChangeValue(field, '299');
        });
    });
    
    describe('quantity field', () => {
        const field = 'quantity';
        it('should render the value correctly', () => {
            expectFieldHasValue(field, props.quantity);
        });

        it('should allow only number, tab and backspace', () => {
            expectOnlyNumber(field);
        });

        it('should update value when onChange is called', () => {
            expectChangeValue(field, '2');
        });
    });
    
    it('should pass the title prop to the Page component', () => {
        expect(wrapper.find('Page').prop('title')).toBe(props.title);
    });
    
    it('should go back to the list when cancel is clicked', () => {
        wrapper.find('.product-form-page__cancel-button').prop('onClick')()
        expect(mockPush).toHaveBeenLastCalledWith('/');
    });
    
    it('should prevent default submit action and call onSubmit when form is submitted', () => {
        const mockEvent = {
            preventDefault: jest.fn()
        };
        wrapper.find('form').prop('onSubmit')(mockEvent);
    
        expect(mockEvent.preventDefault).toHaveBeenCalled();
        expect(handleSubmit).toHaveBeenLastCalledWith({
            name: props.name,
            price: props.price,
            quantity: props.quantity
        });
    });
});