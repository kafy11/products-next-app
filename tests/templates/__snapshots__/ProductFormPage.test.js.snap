// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductFormPage should render correctly 1`] = `
<Page
  title="Product Form"
>
  <form
    className="mb-3"
    onSubmit={[Function]}
  >
    <InputText
      className="product-form-page__name"
      label="Name"
      onChange={[Function]}
      required={true}
      value="Test"
    />
    <InputText
      append=".99"
      className="product-form-page__price"
      label="Price"
      onChange={[Function]}
      onKeyPress={[Function]}
      prepend="$"
      required={true}
      value="10"
    />
    <InputText
      className="product-form-page__quantity"
      label="Quantity"
      onChange={[Function]}
      onKeyPress={[Function]}
      required={true}
      value="1"
    />
    <Button
      submit={true}
    >
      Submit
    </Button>
    <Button
      className="ml-2 product-form-page__cancel-button"
      color="light"
      onClick={[Function]}
    >
      Cancel
    </Button>
  </form>
</Page>
`;
